### ISSUE TITLE
Some valid MD051 header slugs don't work in VS Code Markdown Preview

### ISSUE THREAD

**Issue by redactedscribe** [DavidAnson/vscode-markdownlint#278]
Sometimes I encounter an issue where makdownlint (MDL) will complain with [MD051](https://github.com/DavidAnson/markdownlint/blob/v0.27.0/doc/md051.md) because the auto suggested header slug / anchor offered by VS Code differs from what MDL expects. I'm guessing this is because they both offer differing header slug algorithms: GitHub and whatever VS Code is using.

If I use the header slug considered invalid, the link can be clicked and the section jumped-to in both the editor and in the Markdown Preview. However if I use the valid one, the link will only work for the editor, not in Markdown Preview. I'm not sure what should be fixed and by who, but ideally MDL valid links would work in both cases.

Perhaps MDL could implement the same algorithm that VS Code employs and offer a choice between the two?

The specific case I encountered was this:

```markdown
[A](#-dir-name) <!-- MD051: Valid -->
[B](#dir-name) <!-- MD051: Invalid. Suggested by VS Code -->

## "+ Dir" name
```

Assuming you've implemented the slugification algorithm correctly, it does seem odd to me that a sequence of `"+{space}` equates to `-`: Since the first two characters are stripped (`"+`), I think it would make sense to also strip the leading space, but instead the algorithm treats it as a space between words resulting in `-`.

In any case, what would need changing to have valid MDL header slugs that always work in both the editor and the editor preview?

Thanks.

**Edit:** An issue exists on VS Code's repo regarding [customising the slugify mode algorithm that VS Code uses](https://github.com/microsoft/vscode/issues/106296), but it got no traction. Does this suggest VS Code would need changing and not MDL?

MDL 0.49.0
VS Cod*ium* 1.75.1

**Comment by DavidAnson** [comment]
MD051 attempts to enforce the GitHub algorithm. We can check what GitHub produces by editing a Markdown document and then looking at the Preview tab.

For the Markdown `## "+ Dir" name`, GitHub renders: `#-dir-name`.

Per your example above, that syntax is what MD051 is reporting as valid, so I believe the rule's behavior is correct.

It's true that other syntaxes can work (even in GitHub), but I consider the one it produces to be the "official" correct version.

I don't know that I'd say VS Code's suggestion is wrong here, but markdownlint's current behavior is deliberate.
